Implement basic authentication

Sakif16 opened the pull request:
Implemented a basic authentication using Auth.js and Google provider. The UI is in progress.

Sakif16 committed:
Implemented basic authentication.

Sakif16 committed:
Removed redundant plugin specification

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

orpheus6678 committed:
Fix signIn issue

Navbar.tsx moved from components/ui to components
in Navbar.tsx signIn is now imported from @/auth instead of next-auth/react
Removed accidentally caused redundancies in tailwind.config.ts
Set prettier to use spaces (applied immediately to page, layout.tsx and tailwind.config.ts)
Updated Jest snapshots so tests pass

orpheus6678 reviewed (changes requested):
I dealt with the signIn problem in my last commit. Please modify Navbar.tsx so it would show a profile card (name, profile picture) and logout button upon successful login, and we will proceed onto the next phase (registration page, UI design, etc) soon. 

On a side note, I see that you changed the value of `tailwind.baseColor` in components.json from zinc to slate. It modified a lot of the presets in globals.css. Was this change deliberate?

orpheus6678 committed:
Reconfigure shadcn to not use CSS variables

Warhammer4000 commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/warhammer4000/pr-checks/e250f86b-9864-4f54-95c1-7b6b559f38a9?refs=2bcf23be1cc1b0fc21dcef286d68f092d55f2936&source=prChecksComment)

orpheus6678 committed:
Finished Intro Page and Added vscode-jest

Additionally wrote unit tests for the page
Configured vscode-jest for current workspace

orpheus6678 commented:
I was not the wisest to push an intro page into an authentication feature branch but oh well, too late now

orpheus6678 commented on src/app/page.tsx:
```diff
@@ -1,10 +1,82 @@
+      </div>
+      <Separator className="mx-auto w-[90%] border-2 border-lime-500" />
+
+      <div className="flex flex-col items-center">
+        <h2 className="mb-[3.5rem] mt-[4rem] indent-[.3em] text-2xl uppercase tracking-[.3em]">
+          And more
+        </h2>
+        <Button className="mb-[8rem] uppercase">Get Started</Button>
+      </div>
```
Nah fam

Sakif16 committed:
Initial Login Page Implementation

Implemented Github OAuth and Google session handling

Sakif16 committed:
Merge branch 'feature/authentication' of https://github.com/Learnathon-By-Geeky-Solutions/convicts into feature/authentication

Sakif16 committed:
Enhanced the UI of login page

Created app-specific styles for button components

Sakif16 commented on src/app/login/page.tsx:
```diff
@@ -0,0 +1,127 @@
+            </Button>
+            <Button variant="custom_lime" size="xlg">
+              <Image
+                src="/github_logo.svg"
+                alt="google logo"
+                width={40}
+                height={40}
+              />
```
oopsies :3

orpheus6678 committed:
Implement signup page UI

orpheus6678 committed:
Merge branch 'main' into feature/authentication

orpheus6678 committed:
Add Prisma Adapter

Also sort imports across project

orpheus6678 committed:
Add Email based authentication

Use Nodemailer with Google's SMTP server for login verification mails

orpheus6678 committed:
Add functional first time login page draft

Add username to existing schema
Refactor login page to incorporate first time login state
Validate forms via Zod in server actions

orpheus6678 commented on src/lib/utils.ts:
```diff
@@ -1,5 +1,29 @@
+
+import UserUncheckedCreateInput = Prisma.UserUncheckedCreateInput
+
+export const UserSchema = z.object({
+  id: z.string().cuid().optional(),
+  username: z
+    .string()
+    .min(3)
+    .max(20)
+    .refine(
+      async (username) =>
+        !(await prisma.user.findUnique({ where: { username } })),
+      { message: "Username already taken" },
+    ),
+  name: z.string().min(3).max(64).optional(),
+  email: z.string().email(),
+  image: z.string().optional(),
+  createdAt: z.string().datetime().optional(),
+  updatedAt: z.string().datetime().optional(),
+}) satisfies z.Schema<UserUncheckedCreateInput>
```
@coderabbitai saving the value in a variable would enable caching and avoid database calls? how come?

orpheus6678 committed:
Minor refactoring changes

Moved authorization logic to lib
Fixed prettier import order configuration

orpheus6678 committed:
Upgrade to TailwindCSS v4

orpheus6678 committed:
Tailwind v4 + Refactor code for modularity

orpheus6678 committed:
React Email integration + Initial navbar setup

Created a small email template for email verification
Plus added minor Tailwind v4 upgrades to intro and login pages

orpheus6678 commented on src/ui/input.tsx:
```diff
@@ -0,0 +1,21 @@
+  { className, type, ...props }: React.ComponentProps<"input">,
+  ref?: React.Ref<HTMLInputElement>,
+) => (
+  <input
+    type={type}
+    className={cn(
+      "flex h-9 w-full rounded-md border border-zinc-200 bg-transparent px-3 py-1 text-base shadow-xs transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-zinc-950 placeholder:text-zinc-500 focus-visible:ring-1 focus-visible:ring-zinc-950 focus-visible:outline-hidden disabled:cursor-not-allowed disabled:opacity-50 md:text-sm dark:border-zinc-800 dark:file:te ...
+      className,
+    )}
+    ref={ref}
+    {...props}
+  />
```
@coderabbitai No, outline-none was actually changed to outline-hidden in the newest version of Tailwind.

orpheus6678 committed:
Add verification request and auth error pages

orpheus6678 committed:
SonarCloud coverage analysis setup

orpheus6678 committed:
Add missing projectKey and change workflow name

orpheus6678 committed:
Removed quotes from sonar-project.properties

Restricted actions permissions to read-only

orpheus6678 committed:
Merge branch 'sonarqube' into feature/authentication

orpheus6678 committed:
Graceful error handling in first-login.tsx

utils/UserSchema now does slightly better validation
Merging of OAuth accounts is now automatic
Old SQL migration files purged in favor of a singule succinct initial migration

orpheus6678 committed:
Run prisma generate on CI

orpheus6678 committed:
Add tests for /login

orpheus6678 commented on src/__tests__/login.test.tsx:
```diff
@@ -0,0 +1,62 @@
+import { render, screen } from "@testing-library/react"
+
+import "@testing-library/jest-dom"
+
+import LoginPage from "@/app/login/page"
+
+import { auth } from "@/lib/auth"
+
+jest.mock("@/lib/auth")
```
@coderabbitai im using nextJest in jest.config.js and it is incorporating @swc/jest, but still my transformIgnorePatterns isnt matching.

orpheus6678 committed:
Jest reports unidentified token on Prisma adapter

orpheus6678 commented:
I'm closing this PR and renaming this branch as rf/bug-swc as it is being used as reproduce a certain bug possibly within swc. In case youre wondering, rf stands for reproduction fodder.

orpheus6678 commented:
Stupid me thought I found bugs in library authors code.
Say what you will though, I still think that was very unintuitive.
